File attribute md5 checksum tool

pli888 opened the pull request:
# Pull request for issue: #1032 spreadsheet tool workflow has saving performed MD5 checksum calculation to database

This is a pull request for the following functionalities:

* A tool for updating MD5 checksum attributes for files in a dataset

## How to test?

### Prerequisite

Create a new GitLab variable called `FTP_CONNECTION_URL=https://ftp.cngb.org` for your `dev` and `staging` environments. Ensure this variable is present in `.secrets`.

### `Dev` environment

1. Deploy a `dev` GigaDB by running `./up.sh` and go to the penguin dataset at http://gigadb.gigasciencejournal.com:9170/dataset/100006. Click on the `Files` tab and you should see that `File Attributes` column will have no MD5 checksum values. 

2. Execute the following command to run the tool:
```
$ docker-compose run --rm  test ./protected/yiic files updateMD5FileAttributes --doi=100006
```

3. Now refresh the page at http://gigadb.gigasciencejournal.com:9170/dataset/100006. You should see the `File Attributes` column in the `Files` tab will now have MD5 checksum values 

### `Staging` environment

1. Deploy a staging server in the usual manner by instantiating the AWS infrastructure and running `sd_deploy` on your CI pipeline with code from this PR.

2. Connect to bastion and get `psql` command line access to `gigadb` database:
```
# Connect to bastion
$ ssh -i ~/.ssh/<keyname>.pem centos@<bastion_ip>
# Connect to RDS from bastion
$ export PGPASSWORD='<password>'; psql -U gigadb -d gigadb -h <rds_host_name> -p 5432
```

3. Delete all md5 values for file_attributes belonging to 100006 dataset
```
gigadb=> update file_attributes set value = null where file_id=663 or file_id=664 or file_id=17677 or file_id=17678 or file_id=17679 or file_id=17680 or file_id=17681;
```

4. Check md5 values have been deleted in `file_attributes` table
```
gigadb=> select * from file_attributes where file_id=17681;
```

5. Use md5 checksum tool to update md5 values for 100006 dataset files on your local machine:
```
$ cd ops/infrastructure/envs/staging
$ ansible -i ../../inventories name_bastion_server_staging_<your lowercase IAM role> -a "docker run -e YII_PATH=/var/www/vendor/yiisoft/yii -it registry.gitlab.com/gigascience/forks/<your project prefix>-gigadb-website/production_app:staging ./protected/yiic files updateMD5FileAttributes --doi=100006"
``` 

6. Go back to bastion server to check md5 values have been updated in `file_attributes` table
```
gigadb=> select * from file_attributes where file_id=17681;
```

Try looking at the changes after steps 3 and 5 by browsing https://your_GigaDB_staging_host/dataset/100006 and looking at the Files tab. Unfortunately, you might not see the expected changes due to web page caching by the web site.

## How have functionalities been implemented?

The MD5 values come from a `<dataset_doi>.md5` file which is created by the curators using a simple shell script. The `<dataset_doi>.md5` file is made available from the CNGB FTP server via its URL. This URL varies depending on the DOI, for example the URL for `100006.md5` file is https://ftp.cngb.org/pub/gigadb/pub/10.5524/100001_101000/100006/100006.md5, whilst the `102236.md5` file resides at https://ftp.cngb.org/pub/gigadb/pub/10.5524/102001_103000/102236/102236.md5. Given a DOI, the tool invokes the `actionUpdateMD5FileAttributes($doi)` function in `FilesCommand.php` will determine the URL for the `<dataset_doi>.md5` file.

The tool will download the file using `DownloadService`. This Component class uses the `curl/curl` package which is defined for installation in `composer.lock`, `composer.json.dist` and `ops/packaging/Dockerfile` to perform the downloading. Once downloaded, the tool parses the contents of the file which looks like this:
```
23c3241e6bc362d659a4b589c8d9e01c  ./phylogeny_study_update/Pygoscelis_adeliae.pep.gz
826b699c854cc0f06e982d836410a81b  ./readme.txt
```

Since the tool knows the id of the dataset from its DOI and we can query for the file id using the file name, it is possible to then update the MD5 attribute value for each file in the `file_attributes` table. This is done using a new `updateMd5Checksum($file_id, $md5)` function in the `AdminFileController` class.

## Any issues with implementation?

In order to fetch the `$doi.md5` file for a dataset, URLs are tested until the file is found in the `findDatasetMd5FileUrl($doi)` function in `protected/commands/FilesCommand.php`. This function will eventually run out of directory ranges to test in several years time as there is a limited number of directories corresponding to DOI ranges in the array holding them.

The domain name for the FTP link was hard-coded. This has been removed with a new GitLab variable `ftp_connection_url` which is used in `local.php.dist`.

## Any changes to automated tests?

A new function called `tryToUpdateMD5FileAttributes` has been added into `FilesCommandCest` class. The test will download a `100006.md5` from @pli888's personal FTP server and used to update the empty MD5 file attribute for the 100006 dataset files in the database. @pli888's personal FTP server is used instead of the CNGB FTP server because the former is more stable to allow the new functional test to pass.

Configuration to use @pli888's personal FTP server instead of the CNGB FTP server is defined in `protected/config/test.php` which now has the connection to the `gigadb_testdata` database commented out. This allows the `tryToUpdateMD5FileAttributes` test to execute against the `gigadb` dev database which is the same database that Codeception performs its assertions on as defined in `functional.suite.yml`. This meant that all Behat functional tests in the `protected/tests/functional` directory had to be converted to run against the `gigadb` dev database too. This work mainly involved updating the dataset DOIs and dataset file primary keys.

If the test in `FilesCommandCest` fails then you might need install the `curl/curl` package:
```
$ docker-compose run --rm test composer update curl/curl
```

The Behat `features/admin-file-attributes.feature` acceptance test has been deleted. It is replaced with a Codeception acceptance test now by `tests/acceptance/AdminFileAttributes.feature`. N.B. a couple of tests in `admin-file-attributes.feature` might fail locally but this seems to be a problem with errors in Selenium logs with issues relating to Javascript files downloaded from CDNs. Re-trying the tests should lead to them passing locally and in the CI pipeline.

## Any changes to documentation?

`DATASET_UPLOAD.md` has been updated to include a step describing how to update the MD5 checksum file attributes for a dataset in the database.

`README.md` at the root of the repo has been updated to describe how to quickly restore the `gigadb` dev database using `gigadb.pgdmp` in the `sql` directory. 

## Any technical debt repayment?

The Behat `admin-file-attributes.feature` acceptance test has been deleted and ported into Codeception - see new `tests/acceptance/AdminFileAttributes.feature`. This required updating the `data/dev` CSV files with the Bauhinia and Termitomyces datasets.

There has been a problem with Codeception acceptance tests having to take into account that `dev` data is not restored to its original state before each test is executed. `acceptance.suite.yml` has been updated to fix this problem so that the `dev` database is reset to its starting state as defined by the CSV files in `data/dev`. To speed up the resetting of the database, a `gigadb.pgdmp` file is created by the `up.sh` script using `ops/scripts/make_pgdmp_gigadb.sh` and this database dump file is used for database resetting.

There has been a problem in the `dev` GigaDB website which displays the HTML tags sometimes present in a dataset's title and description instead of parsing them to format the text in these sections on dataset web pages. This has been fixed by updating `csv_yii_migration.js` so that the tags are copied into the database tables without being escaped. This required updating acceptance tests such as `Search.feature` so that tags are not part of the assertions.

`generate_config.sh` has been updated to fix a bug with `.secrets` not containing all required GitLab variables.

Whilst testing all acceptance tests passed in `dev`, I found in http://gigadb.gigasciencejournal.com:9170/dataset/100006, the project image is not displayed because its link is https://assets.gigadb-cdn.net/live/images/projects/genome_10khttps://assets.gigadb-cdn.net/live/images/projects/genome_10k/G10Klogo.jpg. This is caused by `m220126_062259_update_image_location_column_project_tab.php` being executed twice, once by `setup_devdb.sh` and then by `setup_testdb.sh` in `up.sh` since it uses `Yii::app()->db->createCommand` to always execute SQL commands on `gigadb` dev database. This bug breaks an acceptance test in `DatasetView.feature` when run in `dev`. This bug has been fixed now.

pli888 committed:
Create actionUpdateMD5FileAttribute function

A new function called tryToUpdateMD5FileAttribute() has been
created in FilesCommandCest for testing actionUpdateMD5FileAttribute
in the FileCommand class.

pli888 committed:
Use dataset DOI that exists

pli888 committed:
Download $doi.md5 file contents

pli888 committed:
WIP Check remote $doi.md5 file exists

pli888 committed:
Download and parse $doi.md5 file

pli888 committed:
Update md5 checksum file attribute

pli888 committed:
WIP update steps in functional test

pli888 committed:
Remove md5 file attribute values for penguin dataset

pli888 committed:
Update test

pli888 committed:
WIP loop through file md5 values

pli888 committed:
Update file attribute with latest md5 value

pli888 committed:
Create private function to determine dataset md5 file URL

pli888 committed:
Create updateMd5Checksum in AdminFileController

This function is used to update the MD5 value for a file.

pli888 committed:
Add help text

pli888 committed:
Create draft SOP for dataset spreadsheet upload tool

Contains information for using the command line tool to update
MD5 checksum values for all files in a dataset.

pli888 committed:
Clean up code

pli888 committed:
Clean up CSV file

Each row is a dataset record.

pli888 committed:
Add authors for dataset 100056

pli888 committed:
Add submitter and image for dataset 100056

pli888 committed:
Add dataset type metadata

Seems like if you don't add dataset type then the dataset page won't
display the title, dataset type and release date.

pli888 committed:
Add files for dataset 100056

pli888 committed:
Add data for file attributes and dataset log

pli888 committed:
WIP Port file attribute behat tests to Codeception

pli888 committed:
Remove space character from create new file attribute button

The space character looks like a typo and it can confuse file
attribute acceptance tests.

pli888 committed:
Add function to detect create new file attribute button

pli888 committed:
Create function for codeception to detect file attributes table

pli888 committed:
Add steps to detect edit and delete file attribute buttons

pli888 committed:
Add step to detect Save button

pli888 committed:
Add test

pli888 committed:
Add delete last_modified file attribute test

pli888 committed:
Add test to save file after deleting file attribute

pli888 committed:
Add new acceptance test

pli888 committed:
Do not escape HTML tags for data in dataset table

pli888 committed:
Add data for Bauhinia dataset 100245

pli888 committed:
Add missing attribute data

pli888 committed:
Add remaining tests

pli888 committed:
Use constant to hold value of attribute id for md5 checksum

This is to avoid the magic number anti-pattern for 605.

pli888 committed:
Fix broken tests

These tests were broken if they are run one after another. This is
because the tests are all related since the dev database (which
these tests are all run against) is not restored to its starting
state before each test is executed.

pli888 committed:
Add more directory ranges

pli888 committed:
Delete admin-file-attributes.feature

These Behat tests have been replaced by those in
AdminFileAttributes.feature.

pli888 committed:
Add step to update file attribute MD5 checksums

pli888 committed:
Update documentation

pli888 committed:
Fix typo in help text in getHelp() function

rija reviewed (commented):
Hi @pli888,

Both locally and on CI, I have an unrelated acceptance test scenario failing:

```
6) main search function: Query for specific author id
 Test  tests/acceptance/Search.feature:Query for specific author id
 Step  I should see a link "genomic data from adelie penguin (<em>pygoscelis adeliae</em>)." to "/dataset/100006" 
 Fail  No links containing text 'Genomic data from Adelie penguin (<em>Pygoscelis adeliae</em>).' were found in page /search/new?keyword=Lambert%20DM&author_id=3371
```

Could it be related to the changes to the CSV files?

pli888 committed:
Fix Query for specific author id test

pli888 committed:
Fix remaining broken tests

pli888 committed:
Fix checks on text containing HTML tags

HTML markup tags have been removed from text to be checked in
tests. This is because the markup tags are now parsed instead of
being displayed in the web page.

pli888 committed:
Use Db module to clean database between tests

Db module is used to load a database dump to clean up database
between acceptance tests. The sql/gigadb.pgdmp file is created by
make_pgdmp_gigadb.sh script.

pli888 committed:
Update file attributes acceptance test

Makes use of the database cleanup after each acceptance test so
tests have been updated since they are not dependent between each
other.

pli888 committed:
Add documentation to restore dev database using gigadb.pgdmp

pli888 committed:
Fix test

This is because tests are not dependent upon each other since data
in dev database is reset between tests.

pli888 committed:
Fix last 2 tests

The last 2 tests were broken because tests no longer are dependant
upon each other since the dev database is reset after every test.

pli888 commented:
>Could it be related to the changes to the CSV files?

The tests in `Search.feature` were broken because they look for HTML tags, e.g. `<em>` in links which are not present anymore. This problem is fixed now.

pli888 committed:
Remove hardcoded dataset file URL origin

pli888 committed:
Change variable name for HTTPS/FTP link to dataset files

pli888 committed:
Use test.php for running functional test

This is to use https://gigatools.net URL connection to fetch
10006.md5 file to run findDatasetMd5FileUrl test.

pli888 committed:
Display error message if md5 file cannot be found

pli888 committed:
Add ftp_connection_url param

This points to a personal web server with better reliability of
fetching files compared to the CNGB server.

pli888 committed:
Add FTP_CONNECTION_URL configuration for local.php

pli888 committed:
Use curl web client for downloading remote files

Allows HTTP error codes to be monitored.

pli888 committed:
Create DownloadService component service

This is for downloading remote files.

pli888 committed:
Add configuration for reseting database after each test

pli888 committed:
Fix functional tests

pli888 committed:
Fix bug when retrieving name of the attribute

pli888 committed:
Fix functional test

pli888 committed:
WIP use gigadb database for functional tests

pli888 committed:
Fix testItShouldOutputDatasetOnly() test in ApiTest

Problem caused by <em> tags in the title XML element.

pli888 committed:
Fix testItShouldSearchWithKeywordAndOutputDatasetOnly func test

Dataset title contains <em> </em> which creates problems with XML
output as the tags become incorporated as XML tags. These tags
have been removed using strip_tags().

pli888 committed:
Fix tests in AdminFileTest

Had to change file_id so it works with gigadb dev data.

pli888 committed:
Fix functional tests in AdminDatasetMockupActionTest

pli888 committed:
Fix functional tests in AdminDatasetMoveFilesActionTest

pli888 committed:
Fix functional tests in AdminDatasetUpdateActionTest

pli888 committed:
Fix functional tests in FileUploadServiceTest

These tests require the User object for the admin@gigadb.org user.
The User object is obtained by findByAttributes() with the dummy
admin@gigadb.org email address because we don't know its primary
key - see dev/gigadb_user.csv.

pli888 committed:
Fix functional tests in DatasetKeywordsTest

pli888 committed:
Fix use of undefined constant form_params error

pli888 committed:
Fix functional test in AdminDatasetSendInstructionsActionTest

pli888 committed:
Fix testSubmitterCanPerformAction() in AuthorisedDatasetFilesUploadAction

For some reason, the test is complaining that test+336@gigasciencejournal.com
user already exists. Therefore, added ON CONFLICT (email) DO NOTHING
to the SQL command in setUpUserIdentity() function so it does not
try to re-create the user with the email address again.

pli888 committed:
Fix functional tests in AuthorisedDatasetFilesAnnotateAction

pli888 committed:
Fix use of undefined constant form_params error again

pli888 committed:
Delete files

These files are not required anymore since API tests will still be
executed by Behat tests for now.

pli888 committed:
Add curl/curl package

pli888 committed:
Use gigadb.pgdmp to restore gigadb database

The gigadb database is used to provide the data source for both
Behat and Codeception functional tests now. The contents of the
gigadb database is reset to its original state prior to tests
being run.

pli888 committed:
Add test to ensure dummy DOI does not resolve to a md5 file

pli888 committed:
Tidy up code

pli888 committed:
Fix project image link bug

In http://gigadb.gigasciencejournal.com:9170/dataset/100006, the
project image was not being displayed because its link was
https://assets.gigadb-cdn.net/live/images/projects/genome_10khttps://assets.gigadb-cdn.net/live/images/projects/genome_10k/G10Klogo.jpg.
This was caused by m220126_062259_update_image_location_column_project_tab.php
being executed twice, once by setup_devdb.sh and then by setup_testdb.sh
in up.sh since the migration script uses Yii::app()->db->createCommand
to always execute SQL commands on gigadb dev database. This bug was
breaking an acceptance test in DatasetView.feature when run in dev.

m220218_062834_update_url_image_tab.php has also been fixed to so
it can be executed on gigadb and gigadb_testdata databases.

pli888 committed:
Revert changes in gigadb_testdata CSV files

No changes required in gigadb_testdata database because we are
now using gigadb database to execute all functional tests.

pli888 committed:
Tidy up code

pli888 committed:
Add ftp_connection_url parameter

pli888 committed:
Add ftp_connection_url variable into console.php.dist

pli888 committed:
Add conditional to only process lines with content in md5 file

pli888 committed:
Update documentation

rija commented on protected/commands/FilesCommand.php:
```diff
@@ -21,12 +22,86 @@ class FilesCommand extends CConsoleCommand
+                        'dataset_id' => $dataset->id,
+                        'name' => $filename,
+                    ));
+                    $adminFileController->updateMd5Checksum($file->id, $md5_value);
```
Hi @pli888,
 the method ``updateMd5Checksum()`` should be on the  ``File`` class as it is a custom setter for the File [aggregate](https://www.martinfowler.com/bliki/DDD_Aggregate.html), and it should not be in ``AdminFileController`` (which feels wrong anyways to make  a command controller depend on a web controller)
and that line should then read:
```
$file->updateMD5Checksum($md5_value);
```
You then don't need to depend on the ``AdminFileController`` class at all as it seems that's the only use of it.

pli888 replied:
This change makes sense and has been implemented now.

rija commented on protected/components/DownloadService.php:
```diff
@@ -0,0 +1,45 @@
+<?php
+
+/**
+ * Service to provide tokens for password reset functionality
```
The DownloadService's PHPDoc class  description seems wrong to me.

pli888 replied:
Fixed now.

rija commented on protected/components/DownloadService.php:
```diff
@@ -0,0 +1,45 @@
+     */
+    public static function downloadFile(string $url)
+    {
+        $curl = new Curl\Curl();
```
Re the HTTP client, 

Are there some reasons you didn't use [Guzzlehttp](https://packagist.org/packages/guzzlehttp/guzzle) which is used in a lot of places in the codebase (especially in *Service classes) and is already listed in the ``composer.json`` and ``composer.lock`` files?

Furthermore, Guzzlehttp is pretty much the industry standard for modern  PHP as far as http clients are concerned (so if you scan our Composer dependencies, you will see that's the http client used the most and by far), and it supports all of the relevant PSRs which Curl don't.

Also, you wouldn't need to meddle with additional PHP extensions in the Dockerfile by doing away with Curl.

pli888 replied:
I didn't know Guzzlehttp is the industry standard. Guzzlehttp is now used in the code.

rija commented on ops/scripts/generate_config.sh:
```diff
@@ -47,10 +47,10 @@ if ! [ -s ./.secrets ];then
     curl -s --header "PRIVATE-TOKEN: $GITLAB_PRIVATE_TOKEN" "${PROJECT_VARIABLES_URL}?per_page=100&page=2"  > .project_var_raw2
     jq -s 'add' .project_var_raw1 .project_var_raw2 > .project_vars.json
-    cat .project_vars.json | jq --arg ENVIRONMENT $REVIEW_ENV -r '.[] | select( .environment_scope == "dev" ) | select(.key | test("key|ca|pem|cert";"i") | not ) |.key + "=" + .value' > .project_var
+    cat .project_vars.json | jq --arg ENVIRONMENT $GIGADB_ENV -r '.[] | select( .environment_scope == "*" or .environment_scope == "dev" ) | select(.key | test("key|ca|pem|cert";"i") | not ) |.key + "=" + .value' > .project_var
```
Hi @pli888,
The FlysystemCest functional test is failing because the AWS credentials are not written to ``.secrets``, and therefore not available when ``generate_config.sh`` build the  ``web.php`` config file. 
This is because an error on that line. 

It should be:
```
cat .project_vars.json | jq --arg ENVIRONMENT $GIGADB_ENV -r '.[] | select( .environment_scope == "dev" or  .environment_scope == "*") | select(.key | test("private_key|tlsauth|ca|pem|cert";"i") | not ) |.key + "=" + .value'
```
as currently the ``test()`` filter is filtering out AWS credentials because they have ``key`` in their variable name.

kencho51 replied:
Hi @pli888,
The FlysystemCest functional test is failing on my dev also, after applying @rija's fix, the `% ./tests/unit_functional_runner` is passing now.

pli888 replied:
`generate_config.sh` has been updated with the above line of code now.

kencho51 commented on data/dev/attribute.csv:
```diff
@@ -10,7 +10,9 @@ id,attribute_name,definition,model,structured_comment_name,value_syntax,allowed_
 318,sample source,"additional information about where the sample came from, e.g. the particular zoo/avery/lab name","",sample_source,"","",m,"",""
 319,sample contact,the person or institute that can be contacted regarding the sample acquisition,"",sample_contact,"","",m,"",""
 320,IUCN Red List,"The IUCN Red List of Threatened Species, is the world's most comprehensive inventory of the global conservation status of biological species. See wikipedia http://en.wikipedia.org/wiki/IUCN_Red_List.  (Extinct (EX) � No known individuals remaining. Extinct in the Wild (EW) � Known only to survive in captivity, or as a naturalized population outside its historic range. Critica ...
```
Is it better to remove the special character `�` from the test data?

pli888 replied:
Not sure what happened but those special characters have now been removed.

only1chunts replied:
I removed the odd chars from the live version, so maybe that change got propagated to the test data?

only1chunts commented:
@kencho51 , I will attempt to remove those characters from the production version of row 320 "IUCN Red List" as they should not be there! Hopefully, that will allow you to keep the test data better aligned with the live. I say attempt because they are not visible on the html view in the admin page so all I'm doing is copying the text to a plain text editor and then pasting it back in to the admin page then saving it!

rija commented:
Hi @pli888, @kencho51,

I've got all unit, and functional tests passing locally after applying the ``jq`` fix I mentioned in previous comment.
All unit and functional tests are also passing on CI.

Running the acceptance tests, I've got a test failing, locally and on CI.
But the failing test is different whether run locally or on CI. 
Have you had issue like that?

Local failure: 
```
1) a user visit the dataset page: pagination widget is shown when total number of file greater than the page size setting
 Test  tests/acceptance/DatasetView.feature:pagination widget is shown when total number of file greater than the page size setting
 Step  I have set the page size setting to 5 
 Fail  Link or Button or CSS or XPath element with 'Files' was not found.

```

CI failure:
```
1) A curator can manage file attributes in admin file update page: Delete a last modified attribute on admin file update page
 Test  tests/acceptance/AdminFileAttributes.feature:Delete a last modified attribute on admin file update page
 Step  I should not see "last_modified" 
 Fail  Failed asserting that  on page /adminFile/update/id/13973

```

rija commented:
Hi @pli888, @kencho51,

I've got all unit, and functional tests passing locally after applying the ``jq`` fix I mentioned in previous comment.
All unit and functional tests are also passing on CI.

Running the acceptance tests, I've got a test failing, locally and on CI.
Have you had issue like that?

```
1) A curator can manage file attributes in admin file update page: Delete a last modified attribute on admin file update page
 Test  tests/acceptance/AdminFileAttributes.feature:Delete a last modified attribute on admin file update page
 Step  I should not see "last_modified" 
 Fail  Failed asserting that  on page /adminFile/update/id/13973

```

rija commented:
Hi @pli888,

Regarding the main functionality: 

I've followed the instructions in "How to test?" section and the functionality worked as expected  without problem on both my dev environment and my staging environment.

I think I'm done with the review of this PR for now as I've got nothing more to add.

kencho51 commented:
> ```
> 1) A curator can manage file attributes in admin file update page: Delete a last modified attribute on admin file update page
>  Test  tests/acceptance/AdminFileAttributes.feature:Delete a last modified attribute on admin file update page
>  Step  I should not see "last_modified" 
>  Fail  Failed asserting that  on page /adminFile/update/id/13973
> ```

Hi @pli888, @rija ,

This acceptance was also failing on my local dev and gitlab pipeline. But after I tried `./ops/scripts/setup_devdb.sh dev` and then `./ops/scripts/make_pgdmp_gigadb.sh` locally, the test could pass now. While in gitlab pipeline, the `a_gigadb` is finally passing after triggering the pipeline with an empty commit.
And the reason for that, in my suspicion, the `gigadb.pgdmp` did not contain the file attribute for id `13973` in the failed case.

Besides the above observation, all tests are passing locally and also in gitlab pipeline. And by following the `How to test` steps, the `md5 checksum tool` works as expected.

pli888 committed:
Move updateMd5Checksum() into File class

The updateMd5Checksum() function is a custom setter for the File
aggregate, it should not be in AdminFileController.

pli888 committed:
Include file_id when searching for file attribute

Otherwise, findByAttributes() will return the first file attribute
it comes across.

pli888 committed:
Use Guzzle as a web client to download remote files

Any code relating to Curl PHP have been removed.

pli888 committed:
Remove weird characters from attribute id 320

pli888 committed:
Skip acceptance tests that have @javascript tag

Only tests in AdminFileAttributes.feature contain the @javascript
tag. These @javascript tagged tests can sometimes randomly fail.

pli888 committed:
Remove @javascript tags

Some tests do not involve pressing Delete or Save button which
contain javascript. @javascript tags in these tests have been
removed.

pli888 commented:
>Running the acceptance tests, I've got a test failing, locally and on CI.
Have you had issue like that?
```
1) A curator can manage file attributes in admin file update page: Delete a last modified attribute on admin file update page
 Test  tests/acceptance/AdminFileAttributes.feature:Delete a last modified attribute on admin file update page
 Step  I should not see "last_modified" 
 Fail  Failed asserting that  on page /adminFile/update/id/13973
```

I've tried looking into this problem but still don't understand what is causing the problem so I've reused the `@javascript` tag to skip tests in `AdmiinFileAttributes.feature` involving clicking the `Delete` or `Save` button for the time being. I've added a `TODO` line before the above breaking test.

kencho51 reviewed (approved):
Hi @pli888,

All tests are passing locally and in gitlab pipeline, especially after commented out the scenario tagged with `@javascript` for the acceptance test.

And I could use this tool to update the md5 values locally and in staging server as well for file attributes of dataset `100006` as expected in the `How to test` session.

The code looks good to me, so I would like to approve this PR.

rija commented on protected/commands/FilesCommand.php:
```diff
@@ -21,12 +22,86 @@ class FilesCommand extends CConsoleCommand
+    private function findDatasetMd5FileUrl($doi): string
+    {
+        // Directory names representing ranges of dataset DOIs
+        $ranges = ['104001_105000', '103001_104000', '102001_103000', '101001_102000', '100001_101000'];
```
Hi @pli888,
I feel like it's better to define ``$ranges`` as a constant in the ``Dataset`` class: it's a characteristic about one of ``Dataset``'s property (identifier) and we might want to use it elsewhere in the codebase in the future.

pli888 replied:
A `RANGES` constant that is an array of range directory names has been created in the `Dataset` model class.

rija commented on protected/commands/FilesCommand.php:
```diff
@@ -21,12 +22,86 @@ class FilesCommand extends CConsoleCommand
+        foreach ($ranges as $range) {
+            $url = Yii::app()->params['ftp_connection_url']."/pub/gigadb/pub/10.5524/$range/$doi/$doi.md5";
+            // Check URL resolves to a real file
+            $file_exists = @fopen($url, 'r');
```
Hi @pli888,

If we are interested only in checking the existence of the md5 file, it's a bit of bandwidth waste to have to download the file to do that check.
May be it's enough to send a HEAD request to the URL endpoint and check the HTTP response is 200 OK?

If so, there is an example in the function beneath it in the same file that uses the PHP's builtin ``get_headers`` to do something identical.

That said, since I talked about standardising on Guzzle, may be we can use Guzzle in both functions (bonus is that it's easier with Guzzle, as we just call ``head()`` on the web client) (and ``DownloadService`` may be the right place to put that new method)

Also, I think we should move away from using ``fopen`` for URLs, as that PHP function was initially (and still primarily is) for opening local file streams
and because requests to URLs have their own standardised semantics for handling success and failure that are not well represented in ``fopen``'s file orientated interface.

pli888 replied:
A static function called `fileExists` has now been created in `DownloadService` class. It uses Guzzle `head()` to check a remote file exists. Unit tests in `tests/unit/DownloadTest.php` check the functionality of the functions in `DownloadService`.

All tests are passing in my `dev` and `CI` environments.

rija commented:
I confirmed that all feedbacks were implemented, and that all the tests are passing locally and on CI

pli888 committed:
Move ranges variable into Dataset class as a constant

pli888 committed:
Use Guzzle to check dataset md5 file exists

pli888 committed:
Update functional test with dummy dataset doi

pli888 committed:
Create unit test for DownloadService

pli888 committed:
Add catch blocks

pli888 committed:
Replace code with fileExists() from DownloadService

pli888 commented:
@rija The `actionCheckUrls()` function in `FilesCommand` has been updated to use the `DownloadService::fileExists()` function now. All tests are passing in `dev` and `CI`.

rija reviewed (approved):
Hi @pli888,

I can see the change I mentioned in meeting was implemented.
I got all the tests passing locally and CI.
Functionality is working, 
I'm happy to approve this PR
